crowbar_framework/app/models/service_object.rb

```diff
@@ -1190,6 +1193,15 @@ def apply_role(role, inst, in_queue)
           tmprole = RoleObject.find_role_by_name "#{role_name}_remove"
           unless tmprole.nil?
             elem_remove = tmprole.name
+            # save remove intention in #{@bc_name}-databag
+            databag["deployment"][@bc_name]["elements"]["to_remove"] ||= {}
+
+            old_nodes.each do |old_node|
+              @logger.debug "saving #{elem_remove} intention for #{old_node}"
+              databag["deployment"][@bc_name]["elements"]["to_remove"][old_node] ||= []
+              databag["deployment"][@bc_name]["elements"]["to_remove"][old_node] << elem_remove
```

Comment:
Of course it's a different object!  After all we didn't write

    to_remove = databag

But that still doesn't explain how there could be different behaviour.  I challenge you to find a value of `hash` for which

    hash['foo']['bar'] = 'baz'

and

    nested_hash = hash['foo']
    nested_hash['bar'] = 'baz'

behave differently :-)  In both cases a lookup is done on the top-level object, returning another object, and then assignment is done on the second object.